File: web/src/lib/matcher.ts
import {
  MedicineRow,
  findMedicineByBrand,
  findMedicineByBrandPrefix,
  findMedicinesByGeneric,
  findMedicinesByGenericStrengthForm,
  findMedicinesBySalt,
  getAllBrandNames,
  getAllGenericNames,
} from "./db";
import Fuse from "fuse.js";

export interface Alternative {
  brand_name: string;
  generic_name: string;
  salt_composition: string;
  manufacturer: string;
  brand_price: number;
  generic_price: number;
  jan_aushadhi_price: number | null;
  unit_type: string;
  category: string;
  strength: string;
  form: string;
  savings_vs_brand: number;
  savings_pct: number;
  source: string;
  price_available: boolean;
}

export interface MedicineMatch {
  query: string;
  matched_brand: string | null;
  salt_composition: string | null;
  match_type: "exact" | "fuzzy" | "salt" | "none";
  fuzzy_score: number;
  alternatives: Alternative[];
  error: string | null;
}

export function findAlternatives(name: string, topN = 5): MedicineMatch {
  const result: MedicineMatch = {
    query: name,
    matched_brand: null,
    salt_composition: null,
    match_type: "none",
    fuzzy_score: 0,
    alternatives: [],
    error: null,
  };

  if (!name || !name.trim()) {
    result.error = "Empty medicine name";
    return result;
  }

  // 1. Exact brand match
  const exact = findMedicineByBrand(name.trim());
  if (exact) {
    result.matched_brand = exact.brand_name;
    result.salt_composition = exact.salt_composition;
    result.match_type = "exact";
    result.fuzzy_score = 100;
    result.alternatives = getAlternativesBySalt(exact, topN);
    return result;
  }

  // 1b. Prefix brand match (e.g. "Dolo" matches "Dolo 650")
  const prefix = findMedicineByBrandPrefix(name.trim());
  if (prefix) {
    result.matched_brand = prefix.brand_name;
    result.salt_composition = prefix.salt_composition;
    result.match_type = "exact";
    result.fuzzy_score = 95;
    result.alternatives = getAlternativesBySalt(prefix, topN);
    return result;
  }

  // 2. Fuzzy brand match
  const allBrands = getAllBrandNames().map((b) => ({ name: b }));
  const fuseBrand = new Fuse(allBrands, {
    keys: ["name"],
    threshold: 0.4,
    includeScore: true,
  });

  const fuzzyResults = fuseBrand.search(name);
  if (fuzzyResults.length > 0 && fuzzyResults[0].score !== undefined) {
    const matched = fuzzyResults[0].item.name;
    const score = Math.round((1 - fuzzyResults[0].score) * 100);

    if (score >= 60) {
      const dbMed = findMedicineByBrand(matched);
      if (dbMed) {
        result.matched_brand = dbMed.brand_name;
        result.salt_composition = dbMed.salt_composition;
        result.match_type = "fuzzy";
        result.fuzzy_score = score;
        result.alternatives = getAlternativesBySalt(dbMed, topN);
        return result;
      }
    }
  }

  // 3. Generic/salt name match
  const allGenerics = getAllGenericNames().map((g) => ({ name: g }));
  const fuseGeneric = new Fuse(allGenerics, {
    keys: ["name"],
    threshold: 0.4,
    includeScore: true,
  });

  const genericResults = fuseGeneric.search(name);
  if (genericResults.length > 0 && genericResults[0].score !== undefined) {
    const matchedGeneric = genericResults[0].item.name;
    const score = Math.round((1 - genericResults[0].score) * 100);

    if (score >= 60) {
      const meds = findMedicinesByGenericStrengthForm(matchedGeneric, "", "");
      const dbMed =
        meds.length > 0
          ? meds.reduce((a, b) =>
              a.brand_price_per_unit > b.brand_price_per_unit ? a : b
            )
          : undefined;

      if (!dbMed) {
        const broader = findMedicinesByGeneric(matchedGeneric);
        if (broader.length > 0) {
          const best = broader.reduce((a, b) =>
            a.brand_price_per_unit > b.brand_price_per_unit ? a : b
          );
          result.matched_brand = best.brand_name;
          result.salt_composition = best.salt_composition;
          result.match_type = "salt";
          result.fuzzy_score = score;
          result.alternatives = getAlternativesBySalt(best, topN);
          return result;
        }
      }

      if (dbMed) {
        result.matched_brand = dbMed.brand_name;
        result.salt_composition = dbMed.salt_composition;
        result.match_type = "salt";
        result.fuzzy_score = score;
        result.alternatives = getAlternativesBySalt(dbMed, topN);
        return result;
      }
    }
  }

  result.error = `No match found for '${name}'`;
  return result;
}

function getAlternativesBySalt(ref: MedicineRow, topN: number): Alternative[] {
  let sameSalt = findMedicinesByGenericStrengthForm(
    ref.generic_name,
    ref.strength,
    ref.form
  );

  if (sameSalt.length === 0) {
    sameSalt = findMedicinesBySalt(ref.salt_composition, ref.form);
  }

  const refPrice = ref.brand_price_per_unit;

  const alts: Alternative[] = sameSalt.map((med) => {
    const savings = Math.max(0, refPrice - med.brand_price_per_unit);
    const savingsPct = refPrice > 0 ? (savings / refPrice) * 100 : 0;
    return {
      brand_name: med.brand_name,
      generic_name: med.generic_name,
      salt_composition: med.salt_composition,
      manufacturer: med.manufacturer,
      brand_price: med.brand_price_per_unit,
      generic_price: med.generic_price_per_unit,
      jan_aushadhi_price: med.jan_aushadhi_price,
      unit_type: med.unit_type,
      category: med.category,
      strength: med.strength,
      form: med.form,
      savings_vs_brand: Math.round(savings * 100) / 100,
      savings_pct: Math.round(savingsPct * 10) / 10,
      source: "local_db",
      price_available: true,
    };
  });

  // Add generic formulation
  const genericSavings = Math.max(0, refPrice - ref.generic_price_per_unit);
  const genericPct = refPrice > 0 ? (genericSavings / refPrice) * 100 : 0;
  alts.push({
    brand_name: "Generic Formulation",
    generic_name: ref.generic_name,
    salt_composition: ref.salt_composition,
    manufacturer: "Various Manufacturers",
    brand_price: ref.generic_price_per_unit,
    generic_price: ref.generic_price_per_unit,
    jan_aushadhi_price: ref.jan_aushadhi_price,
    unit_type: ref.unit_type,
    category: ref.category,
    strength: ref.strength,
    form: ref.form,
    savings_vs_brand: Math.round(genericSavings * 100) / 100,
    savings_pct: Math.round(genericPct * 10) / 10,
    source: "generic",
    price_available: true,
  });

  alts.sort((a, b) => a.brand_price - b.brand_price);

  const seen = new Set<string>();
  const unique: Alternative[] = [];
  for (const a of alts) {
    const key = a.brand_name.toLowerCase();
    if (!seen.has(key)) {
      seen.add(key);
      unique.push(a);
    }
  }

  return unique.slice(0, topN);
}

export function bulkFindAlternatives(names: string[]): MedicineMatch[] {
  return names.map((n) => findAlternatives(n));
}

File: backend/services/matcher.py
"""
Matcher service — maps extracted medicine names to DB records.
Uses fuzzy matching (RapidFuzz) + exact salt/generic lookup.
"""
from __future__ import annotations

import logging
from typing import List, Optional
from dataclasses import dataclass, field

from rapidfuzz import fuzz, process  # type: ignore
from sqlalchemy.orm import Session

from backend.database.db import Medicine
from backend.services.web_medicine import lookup_web_medicine

logger = logging.getLogger(__name__)

FUZZY_THRESHOLD = 72  # minimum score (0-100) to accept a fuzzy match


@dataclass
class AlternativeResult:
    brand_name: str
    generic_name: str
    salt_composition: str
    manufacturer: str
    brand_price: float
    generic_price: float
    jan_aushadhi_price: Optional[float]
    unit_type: str
    category: str
    strength: str
    form: str
    savings_vs_brand: float = 0.0
    savings_pct: float = 0.0
    match_score: int = 100
    source: str = "local_csv"
    source_urls: List[str] = field(default_factory=list)
    price_available: bool = True


@dataclass
class MedicineMatch:
    query: str                          # original extracted name
    matched_brand: Optional[str] = None  # best DB brand match
    salt_composition: Optional[str] = None
    alternatives: List[AlternativeResult] = field(default_factory=list)
    match_type: str = "none"            # "exact" | "fuzzy" | "salt" | "none"
    fuzzy_score: int = 0
    error: Optional[str] = None


def find_alternatives(medicine_name: str, db: Session, top_n: int = 5) -> MedicineMatch:
    """
    Main entry point.  Given a medicine name string, returns a MedicineMatch
    with ranked cheaper alternatives.
    """
    result = MedicineMatch(query=medicine_name)

    if not medicine_name or not medicine_name.strip():
        result.error = "Empty medicine name"
        return result

    name_lower = medicine_name.strip().lower()

    # 1. Exact brand match (case-insensitive)
    exact = db.query(Medicine).filter(
        Medicine.brand_name.ilike(name_lower)
    ).first()

    if exact:
        result.matched_brand = exact.brand_name
        result.salt_composition = exact.salt_composition
        result.match_type = "exact"
        result.fuzzy_score = 100
        result.alternatives = _get_alternatives_by_salt(exact, db, top_n)
        _append_web_alternative(result, medicine_name)
        return result

    # 2. Fuzzy brand match
    all_brands = db.query(Medicine.brand_name).all()
    brand_list = [b[0] for b in all_brands]

    fuzzy_match = process.extractOne(
        medicine_name,
        brand_list,
        scorer=fuzz.token_set_ratio,
        score_cutoff=FUZZY_THRESHOLD,
    )

    if fuzzy_match:
        matched_name, score, _ = fuzzy_match
        db_med = db.query(Medicine).filter(
            Medicine.brand_name.ilike(matched_name)
        ).first()
        if db_med:
            result.matched_brand = db_med.brand_name
            result.salt_composition = db_med.salt_composition
            result.match_type = "fuzzy"
            result.fuzzy_score = int(score)
            result.alternatives = _get_alternatives_by_salt(db_med, db, top_n)
            _append_web_alternative(result, medicine_name)
            return result

    # 3. Fuzzy generic/salt name match
    all_generics = db.query(Medicine.generic_name, Medicine.salt_composition).all()
    generic_list = [g[0] for g in all_generics]
    salt_list = [g[1] for g in all_generics]

    generic_match = process.extractOne(
        medicine_name,
        generic_list,
        scorer=fuzz.token_set_ratio,
        score_cutoff=FUZZY_THRESHOLD,
    )

    if generic_match:
        matched_generic, score, idx = generic_match
        # find a representative entry
        db_med = db.query(Medicine).filter(
            Medicine.generic_name.ilike(matched_generic)
        ).order_by(Medicine.brand_price_per_unit.desc()).first()
        if db_med:
            result.matched_brand = db_med.brand_name
            result.salt_composition = db_med.salt_composition
            result.match_type = "salt"
            result.fuzzy_score = int(score)
            result.alternatives = _get_alternatives_by_salt(db_med, db, top_n)
            _append_web_alternative(result, medicine_name)
            return result

    result.error = f"No match found for '{medicine_name}'"
    web_match = _find_web_info(medicine_name)
    if web_match:
        return web_match
    return result


def _find_web_info(medicine_name: str) -> MedicineMatch | None:
    web = lookup_web_medicine(medicine_name)
    if not web:
        return None

    result = MedicineMatch(query=medicine_name)
    result.matched_brand = web.name
    result.salt_composition = web.salt_composition or web.generic_name
    result.match_type = "web"
    result.fuzzy_score = 0
    result.error = None
    result.alternatives = [
        AlternativeResult(
            brand_name=web.name,
            generic_name=web.generic_name or web.name,
            salt_composition=web.salt_composition or web.generic_name or web.name,
            manufacturer=web.manufacturer or web.source,
            brand_price=0.0,
            generic_price=0.0,
            jan_aushadhi_price=None,
            unit_type=web.form or "unknown",
            category=web.category or "Web lookup",
            strength=web.strength or "",
            form=web.form or "",
            savings_vs_brand=0.0,
            savings_pct=0.0,
            source=web.source,
            source_urls=web.source_urls,
            price_available=False,
        )
    ]
    return result


def _append_web_alternative(result: MedicineMatch, medicine_name: str) -> None:
    web_match = _find_web_info(medicine_name)
    if not web_match or not web_match.alternatives:
        return

    existing = {alt.brand_name.lower() for alt in result.alternatives}
    for alt in web_match.alternatives:
        key = alt.brand_name.lower()
        if key in existing:
            continue
        alt.source = alt.source or "web"
        alt.price_available = False
        result.alternatives.append(alt)
        existing.add(key)

    if result.match_type != "web":
        result.error = None


def _get_alternatives_by_salt(
    reference: Medicine, db: Session, top_n: int
) -> List[AlternativeResult]:
    """
    Fetch medicines with the same generic and strength/salt, then rank by price.
    Always includes a 'generic formulation' entry.
    """
    same_salt = db.query(Medicine).filter(
        Medicine.generic_name.ilike(reference.generic_name),
        Medicine.strength.ilike(reference.strength),
        Medicine.form.ilike(reference.form),
    ).all()

    if not same_salt:
        same_salt = db.query(Medicine).filter(
            Medicine.salt_composition.ilike(reference.salt_composition),
            Medicine.form.ilike(reference.form),
        ).all()

    ref_price = reference.brand_price_per_unit

    alts: List[AlternativeResult] = []
    for med in same_salt:
        savings = max(0.0, ref_price - med.brand_price_per_unit)
        savings_pct = (savings / ref_price * 100) if ref_price > 0 else 0.0
        alts.append(AlternativeResult(
            brand_name=med.brand_name,
            generic_name=med.generic_name,
            salt_composition=med.salt_composition,
            manufacturer=med.manufacturer,
            brand_price=med.brand_price_per_unit,
            generic_price=med.generic_price_per_unit,
            jan_aushadhi_price=med.jan_aushadhi_price,
            unit_type=med.unit_type,
            category=med.category,
            strength=med.strength,
            form=med.form,
            savings_vs_brand=round(savings, 2),
            savings_pct=round(savings_pct, 1),
        ))

    # Add a standalone "Generic Formulation" entry for the reference salt
    savings_generic = max(0.0, ref_price - reference.generic_price_per_unit)
    savings_pct_generic = (savings_generic / ref_price * 100) if ref_price > 0 else 0.0
    alts.append(AlternativeResult(
        brand_name="Generic Formulation",
        generic_name=reference.generic_name,
        salt_composition=reference.salt_composition,
        manufacturer="Various Manufacturers",
        brand_price=reference.generic_price_per_unit,
        generic_price=reference.generic_price_per_unit,
        jan_aushadhi_price=reference.jan_aushadhi_price,
        unit_type=reference.unit_type,
        category=reference.category,
        strength=reference.strength,
        form=reference.form,
        savings_vs_brand=round(savings_generic, 2),
        savings_pct=round(savings_pct_generic, 1),
    ))

    # Sort cheapest first
    alts.sort(key=lambda a: a.brand_price)

    # Deduplicate by brand name
    seen: set = set()
    unique: List[AlternativeResult] = []
    for a in alts:
        key = a.brand_name.lower()
        if key not in seen:
            seen.add(key)
            unique.append(a)

    return unique[:top_n]


def bulk_find_alternatives(medicine_names: List[str], db: Session) -> List[MedicineMatch]:
    """Process a list of medicine names."""
    return [find_alternatives(name, db) for name in medicine_names]
